api/winget.ts

```diff
@@ -0,0 +1,168 @@
+import got from '../libs/got'
+import { parseDocument } from 'yaml'
+import { basename, extname } from 'path'
+import { version, versionColor } from '../libs/utils'
+import { createBadgenHandler, BadgenError, PathArgs } from '../libs/create-badgen-handler'
+
+const last = <T>(arr: T[]): T => arr[arr.length - 1]
+const rand = <T>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)]
+
+interface Part {
+  number: number
+  other: string
+}
+
+class Version {
+  _source: string
+  _parts: Part[]
+
+  constructor (input: string) {
+    this._source = input
+
+    const parts = input.split('.').map(segment => {
+      return new Version.Part(segment)
+    })
+
+    while (parts.length) {
+      const part = last(parts)
+      if (part.number || part.other) break
+      parts.pop()
+    }
+
+    this._parts = parts
+  }
+
+  get parts() {
+    return this._parts
+  }
+
+  toString() {
+    return this._source
+  }
+
+  static comparator(versionA: Version, versionB: Version): number {
+    let i = 0
+    while (i < versionA.parts.length) {
+      if (i >= versionB.parts.length) break
+
+      const partA = versionA.parts[i]
+      const partB = versionB.parts[i]
+      const result = Version.Part.comparator(partA, partB)
+      if (result) return result
+
+      i += 1
+    }
+
+    if (versionA.parts.length < versionB.parts.length) return -1
+    if (versionA.parts.length > versionB.parts.length) return 1
+    return 0
+  }
+
+  private static Part = class implements Part {
+    _source: string
+    _number: number
+    _other: string
+
+    constructor(input: string) {
+      this._source = input
+
+      const [num, rest] = input.split(/([^\d]+)/)
+      this._number = parseInt(num, 10) || 0
+      this._other = rest
+    }
+
+    get number() {
+      return this._number
+    }
+
+    get other() {
+      return this._other
+    }
+
+    toString() {
+      return this._source
+    }
+
+    static comparator(partA: Part, partB: Part): number {
+      if (partA.number < partB.number) return -1
+      if (partA.number > partB.number) return 1
+      if (partA.other < partB.other) return -1
+      if (partA.other > partB.other) return 1
+      return 0
+    }
+  }
+}
+
+export default createBadgenHandler({
+  title: 'winget',
+  examples: {
+    '/winget/v/GitHub.cli': 'version',
+    '/winget/v/Balena.Etcher': 'version',
+    '/winget/license/Arduino.Arduino': 'license'
+  },
+  handlers: {
+    '/winget/:topic<v|license>/:appId': handler
+  }
+})
+
+async function handler ({ topic, appId }: PathArgs) {
+  const versions = await fetchVersions(appId)
```

Comment:
This request can be moved to `case 'v': {...}`, or we will fetch version twice for license badge

Reply:
Thank you! You are absolutely right. Ah, those joys of late-night coding 🙃